Reduce use of unchecked indexing in parser code

anka-213 opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->
This draft is currently just an example of the kind of changes that would be done to the rest of the parser.

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->
By replacing `spans[idx]` with `spans.get(idx)` calls, we can guarantee that we won't get any Index Out of Bounds panics, since we get back `Option<Span>` where we'll need to actually check that it was valid.

Some of the ones I replaced are very obviously checked, but that means that I can replace a check by a pattern match. And the goal is to remove all unchecked indexing in the parser.

- [x] Fixes #9072 and adds test for it
- Should prevent more errors like #10380 from appearing

## Further ideas:
- Make a new type for non-empty arrays of spans that guarantees that they are non-empty on construction.

# User-Facing Changes
None. Just refactoring and preventing crashes.

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:
-->
- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- [x] `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- [x] `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library
<!--
> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->
No user-facing changes

# More info
The approach is inspired by the [parse don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) idea, where instead of first having a validation step where we check that the data is valid and then later construct data with implicit assumptions about the data, we combine these two into a single step, where we try to construct the data and if the data was invalid, we get a `None` back.

sholderbach commented:
Thanks for diving into this! Changes like this sound good. I think we need to get some of the resident parser experts at the table, as the long term goal should be that the parser state is correct on top of the invalid acccesses getting handled gracefully (which remains a must for stability).

(I smiled at the initial PR description :) )

anka-213 commented:
I don't even know exactly what I changed, but when I refactored the function that crashed before to use my new safer primitives, it resolved the issue in #9072. I did however notice that the new error message  refers to the wrong index, so that's probably related to the cause of the original issue.

As you can see now the first test says "missing f", which is true, but I would have expected it to complain about `e` first. The second test says "missing e", but points at the last argument, `g`, which is very confusing.
```
running 1 test
stdout: 
stderr: Error: nu::parser::missing_positional

  × Missing required positional argument.
   ╭─[]
 1 │ def a [b: bool, c: bool, d: float, e: float, f: float] {}; a true true 1 1
   ·                                                                           ▲
   ·                                                                           ╰── missing f
   ╰────
  help: Usage: a <b> <c> <d> <e> <f>


stdout: 
stderr: Error: nu::parser::missing_positional

  × Missing required positional argument.
   ╭─[]
 1 │ def a [b: bool, c: bool, d: float, e: float, f: float, g: float] {}; a true true 1 1
   ·                                                                                     ▲
   ·                                                                                     ╰── missing e
   ╰────
  help: Usage: a <b> <c> <d> <e> <f> <g>


test tests::test_parser::too_few_arguments ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 528 filtered out; finished in 3.08s
```

Well, at least it doesn't crash anymore and the error messages are technically true.

anka-213 commented:
Oh, I see. I hadn't reran all the other tests. Apparently, what inadvertently fixed the new test caused 95 other tests to fail.

anka-213 commented:
Oh, I figured out part of the original bug, this line https://github.com/nushell/nushell/blob/3a04bd9154fbde1066a9e30eb6b66fa1d7556bc2/crates/nu-parser/src/parser.rs#L968
only checks for `== end`, but in this case we were for some reason looking at two steps from the end. After reading `["true", "true", "1"] at index 2` to get the first float, we try to read ` ["true", "true"] at index 3` to get the second float, which doesn't make sense.

My change ensures that it actually fails that test and skips to the next step, after which we don't get quite a correct error message, but at least a more sensible one. My guess is that there is a bug in `calculate_end_span`, maybe it thinks that the bools are key-value args or something.

jntrnr commented:
For that area of the code - the use that we're using should be correct, and if there are panics, we want to see them so we can fix them. You can't get a span idx beyond the end of the array, so for the cases where you can, that shows an error in the code rather than a problem with not being defensive in how we handle the input.

anka-213 commented:
The actual bug was here
https://github.com/nushell/nushell/blob/3a04bd9154fbde1066a9e30eb6b66fa1d7556bc2/crates/nu-parser/src/parser.rs#L590
It should be
```
 && spans.len() - spans_idx > (signature.required_positional.len() - positional_idx) 
 ```
checking there are fewer remaining positional args than *remaining*  args, rather than fewer positional args than total args.

@jntrnr I see your point, but it would still be preferable for the parser to not crash from a user perspective, so maybe printing a warning in unexpected cases or something like that instead?

anka-213 commented:
Now all the tests passes again. I had a mistake in my translation, where I had accidentally changed `span.0` into `0`

anka-213 commented on crates/nu-parser/src/parser.rs:
```diff
@@ -587,7 +594,7 @@ fn calculate_end_span(
                 spans.len()
             } else if positional_idx < signature.required_positional.len()
-                && spans.len() > (signature.required_positional.len() - positional_idx)
+                && spans.len() - spans_idx > (signature.required_positional.len() - positional_idx)
```
This is the fix for #9072

anka-213 commented on crates/nu-parser/src/parser.rs:
```diff
@@ -3137,24 +3149,25 @@ pub fn parse_var_with_opt_type(
             Expression {
                 expr: Expr::VarDecl(id),
-                span: span(&spans[*spans_idx..*spans_idx + 1]),
+                span: spans.current(),
```
why were we recreating a single span in a complicated way here?

```
span(&spans[*spans_idx..*spans_idx + 1]),
```
is equivalent to
```
spans[*spans_idx]
```
as far as I can tell

anka-213 commented:
This is ready for review now. I have some plans to keep refactoring more of the parser in the same style, but I might as well do that in a new PR, since this is self-contained as is.

anka-213 commented:
Oops, I forgot that subtraction is not a safe operator with unsigned integers. It should work now

jntrnr commented:
@anka-213 - if you found where the bugfix should be, that'd be a good one to create a PR for.

We're still pre-1.0. Users who are using Nushell now are helping us test and get it ready for 1.0 (that's why we put the disclaimer in the README). I get that it's annoying as a user to hit issues, but we want folks to file issues. If it's just a warning, I think some people may not even see it as it might be part of a script they're running.

Not saying I want the parser to crash. I just think we need to be hammering it into shape. Finding issues like the one you found, fixing them, and repeating until we can't find any anymore.

anka-213 commented:
@jntrnr I've extracted the bug fix into #10395 now.

Yeah, I see your point and I have changed the code to as much as possible be a refactor rather than permitting more programs than before (which I accidentally did before). The main difference should be in moving crashes closer to the source, by enforcing invariants on construction, rather than something unrelated crashing later down the line because of the broken invariants.

I would view it as much more of an extension of the "don't use unwrap" guideline than a case of defensive programming. Using `.unwrap()` quietly assumes that there is an invariant that holds that guarantees the validity, which using `a[i]` indexing also does. The goal is to make the invariants more explicit and present in the types rather than implicit and relatedly to make the validity checking explicitly connected to the construction of values and so avoiding "boolean blindness". The refactor simplifies the code in the sense that it reduces the amount of implicit invariant tracking, but also sometimes by removing the need for some code.

I was already from the start trying to avoid changing any behaviour (just writing the exact same checks as before using new primitives), but now I see that I shouldn't view crashes as a thing to avoid as long as they only happen if there is a bug. Which also makes the refactor simpler, since now any cases that were missing from the old checks can be replaced by a call to panic, instead of trying to figure out what could cause it to happen and how to handle it.

anka-213 commented:
Here's a rather silly crash I found by looking at what assumptions indexing operations have and working backwards from that:
```
alias "export alias" = foo
export alias
```
as soon as you type the "s" in the final "alias" it crashes with
```
thread 'main' panicked at 'index out of bounds: the len is 2 but the index is 2', crates/nu-parser/src/parser.rs:241:42
```

Now, this is probably not something a user should ever do, so this is more of an illustration of broken assumptions than a serious bug.

My goal is to make the code less fragile so it's more difficult have incorrect assumptions without noticing.
